### PR TITLE
drivers/mtd/mtd_nandram: Adds virtual NAND Flash simulator

### PR DESCRIPTION
**Pull request by resyfer**
Adds virtual NAND Flash simulator

## Summary

Adds virtual NAND Flash simulator.
Adds log wrapper methods for NAND Flash.
Adds documentation on NAND Flash.

## Impact

Virtual NAND Flash

## Testing

Basic file I/O with littlefs.
`checkpatch.sh` (gives error for multiple data definitions due to array of numbers separated by commas).

### CODE REVIEW

File: Documentation/applications/testing/nand_sim/index.rst [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+=======================
+
+Most NAND flashes share a common interface, specified by the
+`Open NAND Flash Interface (ONFI) <https://www.onfi.org/>`_.
+
+The important part from it, required in this context, is that a NAND Flash is
+divided into a lot of blocks. And each blocks are divided into a lot of
+pages.
+
+Here's the pecular bit. If you want to erase a page, you need to erase
```
Comment (acassis) [inline Documentation/applications/testing/nand_sim/index.rst:22]:
```suggestion
Here's the peculiar bit. If you want to erase a page, you need to erase

**Comment by acassis** [comment]
@resyfer Really nice documentation! Thank you very much for that!

File: Documentation/applications/testing/nand_sim/index.rst [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+These wrapper methods just log the infomation about it being called, and
+directly pass the parameters to the actual upper half (the methods defined
+in ``drivers/mtd/mtd_nand.c``).
+
+Registering Device & Daemon
+===========================
+
+This wrapper is then registered using ``register_mtddriver``, and this
+whole thing is converted to be a daemon, so that the device can keep running
+in the baackground.
```
Comment (xiaoxiang781216) [inline Documentation/applications/testing/nand_sim/index.rst:152]:
```suggestion
in the background.
```

**Comment by resyfer** [comment]
@xiaoxiang781216 Hi, can you please help me with this comma-separated-array-definition being caught by nxstyle? Thank you!

The line is [this](https://github.com/apache/nuttx/pull/11806/files#diff-5ce46229f897184958595ddc83adc65fe0b7ab5577265b38b4051c3dfdb2c4cbR210).

---

> @resyfer Really nice documentation! Thank you very much for that!

Thanks! It was a pretty fun learning processing.

File: drivers/mtd/mtd_nandram.c [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+ *
+ ****************************************************************************/
+
+static void nand_ram_storage_init(void)
+{
+  int i;
+
+  /* Hard coded array for bad block indexes */
+
+  int nand_ram_rand_bad_blk_indx[] = {4, 14, 19, 21, 28, 30, 107, 108, 164,
```
Comment (xiaoxiang781216) [inline drivers/mtd/mtd_nandram.c:210]: try this:
```suggestion
  int nand_ram_rand_bad_blk_indx[] =
  {
    4, 14, 19, 21, 28, 30, 107, 108, 164,
    ...
  };
```
BTW, it's better to move the const array to global section

Comment (resyfer) [reply drivers/mtd/mtd_nandram.c:210]: I'll give this a try thanks!

**Comment by resyfer** [comment]
I don't understand why the arm tests are failing. I haven't touched arch specific code. Is it because of a commit from rebase?

**Comment by xiaoxiang781216** [comment]
> I don't understand why the arm tests are failing. I haven't touched arch specific code. Is it because of a commit from rebase?

Fix here, please wait a moment: https://github.com/apache/nuttx/pull/11833

**Comment by xiaoxiang781216** [comment]
> I don't understand why the arm tests are failing. I haven't touched arch specific code. Is it because of a commit from rebase?

the fix was merged, please rebase again @resyfer .

**Review by acassis — APPROVED** [review]

**Comment by resyfer** [comment]
Thanks for the review!

**Comment by acassis** [comment]
Thank you for your great contribution!

**Comment by xiaoxiang781216** [comment]
@acassis @resyfer this patch has ci break, please fix it to unblock ci:
https://github.com/apache/nuttx/actions/runs/11459507141/job/31884231071?pr=14449

**Pull request merged**